design/boto-design.md

```diff
@@ -0,0 +1,101 @@
+
+## Proposed Design
+The botwo package provides a boto3 like client with two boto3 clients.
+It handles several things for the user:
+* Holds two boto clients and routes the requests between them by the bucket/prefix configuration.
+* Maps 1:1 with the boto AWS service API. All botwo client AWS functions will be generated on the fly according to boto3 client AWS functions.
+* Holds special treatment for two buckets functions like `upload_part_copy`. If more two buckets functions will be added to s3 in the future, we will need to implement them as well.
+* It has more functionality like the event-system that will be implemented by using one of its clients (the default one).
+
+## Alternatives Considered
+
+### Route the requests to lakefs endpoint by the prefix
+```
+s3 = boto3.client('s3')
+if bucket == "bucket_a" and prefix == "my_path":
+  client.endpoint_url  = 'https://lakefs.example.com'
+```
+Does not work because once you initialize boto client with parameters you can’t change them in run time.
+
+### Method Overriding
+
+Monkey patching / overriding base class functions in boto3 with inheritance.
+Does not work because Boto3 functions are generated on the fly from service definitions, so there is no static put_object function to inherit from.
+
+### Creating custom resource-class using the events system
+
+When extending a resource class using the [events system][Extensibility guide], there is no inheritance relationship between the base class and the custom class. It results in a base class enriched with the functions and attributes of the custom class.
+It allows only to replace the bucket and prefix names.
+
+### Route all the requests to lakefs server and implement the bucket configuration there
+It can result in server load because requests like put_object can call directly to s3.
+
+### Wrap boto core fs
+Possible but complex because we will need to implement boto core interface which has a lot of functionality.
+
+## Usage
+
+Current Usage (only with s3):
+```
+import boto3
+
+s3 = boto3.client('s3')
+
+s3.get_object(Bucket="test-bucket", Key="test/object.txt") # in a lot of places
+```
+
+Future Usage: change in every place in the code where there is boto client initialized.
+```
+import boto3
+import botwo
+
+client1 = boto3.client('s3')
+client2 = boto3.client('s3', endpoint_url='https://lakefs.example.com')
+
+mapping = [('bucket-a/*', {'endpoint': 'lakefs', 'bucket': 'my-repo', 'prefix': 'dev/'}),
+('bucket-b/foo', {'endpoint': 'lakefs', 'bucket': 'bucket-b', 'prefix': 'master/'}),
+# default
+('*', 's3')]
+
+s3 = botwo.client({"s3": client1, "lakefs": client2}, mapping)
+
+s3.get_object(Bucket="test-bucket", Key="test/object.txt") # stays the same
+```
+
+Wildcard matching- every time s3 function is called, it will pass on the mapping prefix list by order and stop if there is a match.
```

Comment:
```suggestion
Wildcard matching- every time s3 function is called, it will go over the mapping prefix list by order and stop if there is a match.
```

Reply:
Done